lib/EnzymeTestUtils/src/generate_tangent.jl

```diff
@@ -17,14 +17,20 @@ map_fields_recursive(f, x::T...) where {T<:AbstractFloat} = f(x...)
 map_fields_recursive(f, x::Array{<:Number}...) = f(x...)
 
 rand_tangent(x) = rand_tangent(Random.default_rng(), x)
-rand_tangent(rng, x) = map_fields_recursive(Base.Fix1(rand_tangent, rng), x)
-# make numbers prettier sometimes when errors are printed.
-rand_tangent(rng, ::T) where {T<:AbstractFloat} = rand(rng, -9:T(0.01):9)
-rand_tangent(rng, x::T) where {T<:Array{<:Number}} = rand_tangent.(rng, x)
+function rand_tangent(rng, x)
+    v, from_vec = to_vec(x)
+    T = eltype(v)
+    # make numbers prettier sometimes when errors are printed.
+    v_new = rand(rng, -9:T(0.01):9, length(v))
+    return from_vec(v_new)
+end
 
-zero_tangent(x) = map_fields_recursive(zero_tangent, x)
-zero_tangent(::T) where {T<:AbstractFloat} = zero(T)
-zero_tangent(x::T) where {T<:Array{<:Number}} = zero_tangent.(x)
+# differs from Enzyme.make_zero primarily in that reshaped Arrays in the argument will share
+# the same memory in the output.
+function zero_tangent(x)
```

Comment:
So technically we could still implement this with make zero.

Specificially if we passed in say a tuple (output, input) if out and in aliased make_zero would make the results alias as well.

Alternatively we could prepopulate the seed dict.

Reply:
> Specificially if we passed in say a tuple (output, input) if out and in aliased make_zero would make the results alias as well.

This doesn't work in the case of a reshaped array of floats, which share the same memory but are treated by Julia as non-equivalent. The difference between `make_zero` and `zero_tangent` is this:

```julia
julia> x1 = randn(3, 1);

julia> x2 = vec(x1);

julia> x = (x1, x2);

julia> z1 = Enzyme.make_zero(x)
([0.0; 0.0; 0.0;;], [0.0, 0.0, 0.0])

julia> z2 = EnzymeTestUtils.zero_tangent(x)
([0.0; 0.0; 0.0;;], [0.0, 0.0, 0.0])

julia> z1[1][1] = 1
1

julia> z1
([1.0; 0.0; 0.0;;], [0.0, 0.0, 0.0])

julia> z2[1][1] = 1
1

julia> z2
([1.0; 0.0; 0.0;;], [1.0, 0.0, 0.0])
```

> Alternatively we could prepopulate the seed dict.

Wouldn't this add more code complexity than what we currently do? i.e. we already have a `to_vec` that knows how to handle reshaped float arrays, and we use that for all tangent generation, both for FiniteDifferences, but also for constructing zero and random tangents.

Reply:
Yeah fair, was a suggestion if it worked and simplifies but otherwise LGTM

Reply:
Though wait would testutils work on a reshape call?